src/v/cloud_storage/remote_partition.cc

```diff
@@ -229,90 +323,51 @@ class partition_record_batch_reader_impl final
 
 private:
     /// Return or evict currently referenced reader
-    ///
-    /// If the reader belongs to existing segment it will be returned
-    /// back to the materialized segment.
-    /// If it belongs to evicted segment the reader itself wil be evicted.
     void dispose_current_reader() {
-        if (!_it.is_invalidated()) {
-            _partition->return_reader(std::move(_reader), _it->second);
-        } else {
-            // The segment was evicted or replaced by the compacted
-            // segment. In this case we can't return the reader back
-            // to the materialized segment and have to remove it.
-            _partition->materialized().evict_reader(std::move(_reader));
+        if (_reader) {
+            _partition->return_reader(std::move(_reader));
```

Comment:
not sure but it is probably safer to explicitly set `_reader` to nullptr afterwards

Reply:
Looks like this is the case: https://stackoverflow.com/questions/24061767/is-unique-ptr-guaranteed-to-store-nullptr-after-move 👍